libbpf-rs/src/query.rs

```diff
@@ -98,101 +98,305 @@ fn name_arr_to_string(a: &[c_char], default: &str) -> String {
+    insn_off: u32,
+    type_id: u32,
+}
+
+#[derive(Debug, Clone, Default)]
+#[repr(C)]
+pub struct Tag([u8; 8]);
+
 /// Information about a BPF program
 #[derive(Debug)]
 pub struct ProgramInfo {
     pub name: String,
     pub ty: ProgramType,
-    pub tag: [u8; 8],
+    pub tag: Tag,
     pub id: u32,
-    pub jited_prog_len: u32,
-    pub xlated_prog_len: u32,
-    pub jited_prog_insns: u64,
-    pub xlated_prog_insns: u64,
+    pub jited_prog_insns: Vec<u8>,
+    pub xlated_prog_insns: Vec<u8>,
     /// Duration since system boot
     pub load_time: Duration,
     pub created_by_uid: u32,
-    pub nr_map_ids: u32,
-    pub map_ids: u64,
+    pub map_ids: Vec<u32>,
     pub ifindex: u32,
     pub gpl_compatible: bool,
     pub netns_dev: u64,
     pub netns_ino: u64,
-    pub nr_jited_ksyms: u32,
-    pub nr_jited_func_lens: u32,
-    pub jited_ksyms: u64,
-    pub jited_func_lens: u64,
+    pub jited_ksyms: Vec<*const c_void>,
+    pub jited_func_lens: Vec<u32>,
     pub btf_id: u32,
     pub func_info_rec_size: u32,
-    pub func_info: u64,
-    pub nr_func_info: u32,
-    pub nr_line_info: u32,
-    pub line_info: u64,
-    pub jited_line_info: u64,
-    pub nr_jited_line_info: u32,
+    pub func_info: Vec<FuncInfo>,
+    pub line_info: Vec<LineInfo>,
+    pub jited_line_info: Vec<*const c_void>,
     pub line_info_rec_size: u32,
     pub jited_line_info_rec_size: u32,
-    pub nr_prog_tags: u32,
-    pub prog_tags: u64,
+    pub prog_tags: Vec<Tag>,
     pub run_time_ns: u64,
     pub run_cnt: u64,
 }
 
+#[derive(Default, Debug)]
+pub struct ProgInfoIter {
+    cur_id: u32,
+    opts: ProgInfoQueryOptions,
+}
+
+#[derive(Clone, Default, Debug)]
+pub struct ProgInfoQueryOptions {
+    include_xlated_prog_insns: bool,
+    include_jited_prog_insns: bool,
+    include_map_ids: bool,
+    include_line_info: bool,
+    include_func_info: bool,
+    include_jited_line_info: bool,
+    include_jited_func_lens: bool,
+    include_prog_tags: bool,
+    include_jited_ksyms: bool,
+}
+macro_rules! include_field_setter {
+    ($name:ident) => {
+        pub fn $name(mut self, $name: bool) -> Self {
+            self.$name = $name;
+            self
+        }
+    };
+}
+
+impl ProgInfoIter {
+    pub fn with_query_opts(opts: ProgInfoQueryOptions) -> Self {
+        Self{opts: opts, ..Self::default()}
+    }
+}
+
+impl ProgInfoQueryOptions {
+    include_field_setter!(include_xlated_prog_insns);
+    include_field_setter!(include_jited_prog_insns);
+    include_field_setter!(include_map_ids);
+    include_field_setter!(include_line_info);
+    include_field_setter!(include_func_info);
+    include_field_setter!(include_jited_line_info);
+    include_field_setter!(include_jited_func_lens);
+    include_field_setter!(include_prog_tags);
+    include_field_setter!(include_jited_ksyms);
+}
+
+macro_rules! update_vec_field {
+    ($item:expr, $flag:expr, $field_and_vec:ident, $len_field:ident, $default:expr, $result_ready:expr) => {
```

Comment:
Have you tried expressing this as a function? It seems to me as if mutable references to objects could be passed in and that would suffice?